llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp

```diff
@@ -148,17 +154,44 @@ static bool canUsePressureDiffs(const SUnit &SU) {
   return true;
 }
 
-static void getRegisterPressures(bool AtTop,
-                                 const RegPressureTracker &RPTracker, SUnit *SU,
-                                 std::vector<unsigned> &Pressure,
-                                 std::vector<unsigned> &MaxPressure) {
+static void getRegisterPressures(
+    bool AtTop, const RegPressureTracker &RPTracker, SUnit *SU,
+    std::vector<unsigned> &Pressure, std::vector<unsigned> &MaxPressure,
+    GCNDownwardRPTracker &DownwardTracker, GCNUpwardRPTracker &UpwardTracker,
+    ScheduleDAGMI *DAG, const SIRegisterInfo *SRI) {
   // getDownwardPressure() and getUpwardPressure() make temporary changes to
   // the tracker, so we need to pass those function a non-const copy.
   RegPressureTracker &TempTracker = const_cast<RegPressureTracker &>(RPTracker);
-  if (AtTop)
-    TempTracker.getDownwardPressure(SU->getInstr(), Pressure, MaxPressure);
-  else
-    TempTracker.getUpwardPressure(SU->getInstr(), Pressure, MaxPressure);
+  if (!GCNTrackers) {
+    AtTop
+        ? TempTracker.getDownwardPressure(SU->getInstr(), Pressure, MaxPressure)
+        : TempTracker.getUpwardPressure(SU->getInstr(), Pressure, MaxPressure);
+
+    return;
+  }
+
+  // GCNTrackers
+  Pressure.resize(4, 0);
+  MachineInstr *MI = SU->getInstr();
+  if (AtTop) {
+    GCNDownwardRPTracker TempDownwardTracker(DownwardTracker);
+    TempDownwardTracker.bumpDownwardPressure(MI, SRI);
+    Pressure[AMDGPU::RegisterPressureSets::SReg_32] =
+        TempDownwardTracker.getPressure().getSGPRNum();
+    Pressure[AMDGPU::RegisterPressureSets::VGPR_32] =
+        TempDownwardTracker.getPressure().getArchVGPRNum();
+    Pressure[AMDGPU::RegisterPressureSets::AGPR_32] =
+        TempDownwardTracker.getPressure().getAGPRNum();
+  } else {
+    GCNUpwardRPTracker TempUpwardTracker(UpwardTracker);
```

Comment:
So this is the cornerstone of you change. You're keeping two trackers consistent with the order of the scheduling and at every candidate you need to find its impact on the pressure. My concerns are below.

You're making a full copy of tracker object for every candidate at each step. This is not cheap as you're making a deep copy of DenseMap for live register set in every tracker just to throw it away immediately with no use. This would be ok for the start and would be ok if it made the program as simple as:
    temp_tracker(consistent_tracker)
    temp_tracker.advance/recede
    temp_tracker.getPressure
    get rid of temp_tracker
But you've created tons of code just to accomplish above. I believe the reason is that LIS isn't consistent with the incoming candidate.

bumpDownwardPressure it's a harder task with non-consistent LIS. Can you do it as a _const_ method of the GCNDownwardTracker? This means that it should not modify the state of the tracker in anyway and preferably with no extra data member to the object itself (except mutables if it's required for some caching). This way it wouldn't require DenseMap copy and make the intention if bumpDownwardPressure clear: It works on a tentative schedule with non-updated LIS.

Upward direction is different. GCNUpwardTracker is created in mind to be used in the situation when LIS isn't updated. The only place where LIS is used by the tracker is the determination of the live mask for register uses. It just reads the mask from the non-updated LIS and the rationale for this is that the defs should dominate uses for any valid schedule and the mask from the non-updated LIS should work. In other words GCNUpwardTracker  should be able to be used like the snippet above without additional work.

Reply:
Unfortunately, it is not this easy for the upwardRPTracker

The method the upwardTracker uses to determine the use lane masks adds lanemasks from other subranges unrelated to the the subreg use in the instruction.

On the other hand, bumpUpwardPressure uses RegOpers.collect which simply uses TRI.getSubRegIndexLaneMask(SubRegIdx) to get the lane masks.

In short, recede and bumpupward are logically inconsistent for this usecase, and I would prefer to not change recede. This would result in introducing behavioral changes when trackers are disabled.

Reply:
This is coming from a real example which has code like

```
%0:vreg_64 = IMPLICIT_DEF 
%1:vgpr_32 = IMPLICIT_DEF, implicit %0.sub0
%2:vgpr_32 = IMPLICIT_DEF, implicit %0.sub1
...
```

After bottomup scheduling the instruction defining `%2`, we'll speculate the RP for the instruction defining `%1`. When speculating using recede, we use `getLiveLaneMask` to calculate the mask of `%0.sub0`. To do so, we iterate over the `subranges()` of `%0` and find both `sub0` and `sub1` to be live at the position. Thus we inaccurately calculate the usemask of `%0.sub0` as 0xF (for both `sub0` and `sub1`). This is because `%0.sub1` is live at the index for the instruction defining `%1`. 

Whereas `getSubRegIndexLaneMask` (used by `bumpUpwardPressure`) for the `sub0` subregidx just returns 0xC as the mask.

For this small example, there will be no problem. But with wide regs with more subregs, these inaccuracies will effect scheduling. This doesn't impact RP calculations for whole regions with consistent `LIS` ordering because the use mask inaccuracies don't effect the delta between previous use mask and new use mask.

Reply:
After (in bottom-up order) the instruction defining %2 we have %0:1100b live for %0.sub1. After we speculate over instruction defining %1, we get %0:1111b live indeed because now sub0 and sub1 are live. But when we calculate pressure we're accounting for a mask increase from %0:1100b to %0:1111b that is for a 0011b difference and this seems correct.

It seems though that it would be more correct if the mask found by getLiveLaneMask would be ANDed with the mask of the actually used subreg in the instruction. This way it can allow to find live masks for a complicated cases like sub1_sub0 subregs (it can have sub0, sub1 or sub1 and sub0 lives) but not wider than actually used subreg mask.

Reply:
Yes you are correct, there is no issue with this example. I simplified out the problem.

Please consider

```
---
name:            pdiff
tracksRegLiveness: true
body:             |
  bb.0:
  %49:areg_1024 = IMPLICIT_DEF
  undef %63.sub0:vreg_128 = COPY %49.sub0:areg_1024
  %63.sub1:vreg_128 = COPY %49.sub1:areg_1024
  %63.sub2:vreg_128 = COPY %49.sub2:areg_1024
  %63.sub3:vreg_128 = COPY %49.sub3:areg_1024
  
  S_ENDPGM 0, amdgpu_allvgprs
...
```


When speculating the first bottom-up instruction, we will calculate the RP from each of the COPYs. If we do so using recede, we will look at the live lanes of the register based on the LIS ordering (`collectVirtualRegUses` -> `getLiveLaneMask`). For the instruction defining undef %63.sub0, %49 has a LaneMask of 0xFF (all four lanes), The RPTracker would then determine this single lane copy uses 4 VGPRs. Similar problems for the instructions defining sub1 and sub2. Moreover, after scheduling an instruction bottomup, we will still have incorrect difference based RP calculations. bumpUpwardPressure calculates the usemask by doing something similar to what you suggest -- it will calculate the lanemask directly from the subreg being used by the instruction. As a result, it doesn't have these issues.

recede really doesn't work as is for this usecase. As you mentioned, we can adjust it to work for speculation, but I would prefer to contain the changes under the optional flag if possible.